File: Week_03/G20200343030632/Leetcode_78_632.java
import java.util.LinkedList;
import java.util.List;

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();
        helper(nums, 0, new LinkedList<>(), res);
        return res;


    }

    private void helper(int[] nums, int start, List<Integer> curpath, List<List<Integer>> res) {
        res.add(new LinkedList<Integer>(curpath));
        for (int i = start; i < nums.length; i++) {
            curpath.add(nums[i]);
            helper(nums, i + 1, curpath, res);
            curpath.remove(curpath.size() - 1);
        }
    }
}

File: Week_03/G20200343030632/Leetcode_78_632.py
class Solution:
    def subsets(self, nums):
        res = []
        self.helper(0, nums, [], res)
        return res
    def helper(self, nums, curpath, res):
        res.append(curpath)
        for i in range(len(nums)):
            self.helper(nums[i + 1:], curpath + [nums[i]], res)
    